### PR TITLE
Wochoice

### PR DESCRIPTION
**Pull request by Tragicus**
##### Motivation for this change

Adds a proof of Zorn's lemma, Hausdorff maximal principle and the well ordering principle.
Contains a patch of contra that we may want to extract.

##### Checklist

- [ ] added corresponding entries in `CHANGELOG_UNRELEASED.md`

- [ ] added corresponding documentation in the headers

Reference: [How to document](https://github.com/math-comp/math-comp/wiki/How-to-document)

##### Reminder to reviewers

- Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-creating-and-review-PRs)
- Put a milestone if possible
- Check labels

### CODE REVIEW

**Comment by affeldt-aist** [comment]
> Contains a patch of contra that we may want to extract.

Do you mean that it ought to be a different PR? (That may help the review process.)

**Comment by Tragicus** [comment]
I would not go so far as "ought to", the fix is quite trivial, but if you would prefer a different PR, I can do it.

**Comment by affeldt-aist** [comment]
> the fix is quite trivial

Ok. Since there is a number of definitions that are added to `contra.v` and many canonicals, it didn't look trivial to me ^_^.

> if you would prefer a different PR, I can do it.

I think that it is better (different concerns, two commits are maybe better also for the git log).

**Comment by CohenCyril** [comment]
@Tragicus could you rephrase your commit message so as to make
```
Co-Authored-By: Georges Gonthier <georges.gonthier@inria.fr>
```
a single line separated from the title by a white line, and verbatim (otherwise github does not pick it up)

**Comment by proux01** [comment]
Argh, I accidentally merged this instead of https://github.com/math-comp/analysis/pull/1183 and I don't have push right on master to revert that, could someone do it, maybe @CohenCyril  ?

**Comment by proux01** [comment]
Done with @affeldt-aist 
@Tragicus could you please reopen the PR, sorry again.

**Comment by Tragicus** [comment]
Can I also rebase and clean up now?

**Comment by proux01** [comment]
Sure, thanks

**Comment by Tragicus** [comment]
Reopened at https://github.com/math-comp/analysis/pull/1198
